Step4 - 블랙잭 (배팅)

gmltkddl1 opened the pull request:
이전 피드백 중 인터페이스 사용한 이유는 이전 수업에서 역할에 대해 수업을 들어서 활욯해보고 싶어서 그렇게 만들었습니다.
이 경우에는 추상 클래스가 더 적합한거 같아 변경하였습니다.

BlackJackGameResult가 데어터 클래스가 아닌 이유는 equals와 같은 데이터 클래스의 연산이 불필요하였기 때문입니다.

gmltkddl1 committed:
refactor: step3 블랙잭(딜러)

인터페이스를 추상 클래스로 변경
딜러 카드 뽑기 로직 추상화 레벨 변경
size -> count

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '플레이어는 게임이 시작할때 배팅 금액을 정한다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '플레이어는 이기면 배팅 금액 만큼 이익을 얻는다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '플레이어는 이기면 배팅 금액 만큼 이익을 얻는다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '플레이어는 비기면 이익의 변화가 없다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '카드의 합이 블랙잭(21)으로 이기면 배팅금액의 1.5배를 얻는다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '딜러는 이기면 상대의 판돈 만큼 이익을 얻는다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '딜러는 지면 상대의 판돈 만큼 손해를 본다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '비기면 이익에 변화가 없다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

요구사항 작성 '승리한 플레이어에게 이익을 판돈만큼 증가시키고 패배한 플레이어의 이익을 판돈만큼 감소시킨다'와 '딜러가 아닌 플레이어가 블랙잭으로 승리하면 1.5배의 판돈을 지급한다' 구현

gmltkddl1 committed:
feat: step4 블랙잭(배팅)

메인 함수 구현

vsh123 commented on src/main/kotlin/blackjack/view/BlackJackInputView.kt:
```diff
@@ -25,13 +26,21 @@ object BlackJackInputView {
+        val bets = mutableListOf<Int>()
+        playerNames.forEach {
+            println("${it}의 배팅 금액은?")
+            bets.add(readln().toInt())
```
배팅 금액을 입력받을때마다 Player를 만들어내는 방식으로 변경해보면 어떨까요?

vsh123 commented on src/main/kotlin/blackjack/view/BlackJackInputView.kt:
```diff
@@ -25,13 +26,21 @@ object BlackJackInputView {
-        print("딜러 카드: ")
-        println(blackJackDealer.blackJackPlayerCards.cards.map { "${it.number.name} ${it.shape.name}" }.joinToString(","))
+    fun getPlayerBets(playerNames: List<String>): Map<String, Int> {
+        val bets = mutableListOf<Int>()
```
mutableListOf 없이 할 수 있는 방법은 무엇이 있을까요?

대표적으로는 [map](https://kotlinlang.org/docs/collection-transformations.html)이 생각나네요!

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackPlayer.kt:
```diff
@@ -1,11 +1,18 @@
+abstract class BlackJackPlayer(
+    val name: String,
+    val blackJackPlayerCards: BlackJackPlayerCards,
+    var profit: Int,
```
var 프로퍼티가 외부에 setter와 함께 노출이 된 것 같네요!! 외부에서는 set이 불가하도록 만들어보면 어떨까요?

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackNormalPlayer.kt:
```diff
@@ -1,26 +1,44 @@
+    name: String,
+    blackJackPlayerCards: BlackJackPlayerCards,
+    profit: Int = 0,
+    val bet: Int = 0,
```
금액이라는 걸 명확히 해주는 네이밍을 고려해보면 어떨까요?

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackPlayer.kt:
```diff
@@ -1,11 +1,18 @@
 
-interface BlackJackPlayer {
-    fun drawCard(blackJackDeck: BlackJackDeck): Boolean
+abstract class BlackJackPlayer(
```
> 이 경우에는 추상 클래스가 더 적합한거 같아 변경하였습니다.

어떠한 점에서 추상 클래스가 더 적합하다고 생각하셨을까요~?

gmltkddl1 replied:
제가 아는바에 의하면 추상클래스는 interface + property라고 알고 있습니다.
즉 공유하는 property가 있다면 추상클래스가 더 적합하다고 생각합니다. 
이 경우에는 딜러와 플레이어는 각각 게임을 진행하는 역할과 property 모두를 공유하고 있다고 생각하여
추상클래스가 더 적합하다고 생각했습니다

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackPlayResult.kt:
```diff
@@ -7,13 +7,16 @@ enum class BlackJackPlayResult(val resultString: String) {
-            myScore: Int,
-            opponentScore: Int,
+            normalPlayerScore: Int,
+            DealerScore: Int,
```
```suggestion
            dealerScore: Int,
```
으로 수정해주면 좋을 것 같네요! :)

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackGame.kt:
```diff
@@ -12,8 +12,33 @@ class BlackJackGame(
             }
-        val winCount = playerResults.filter { it.result == BlackJackPlayResult.LOSE }.size
-        val loseCount = playerResults.filter { it.result == BlackJackPlayResult.WIN }.size
+        val winCount = playerResults.filter { it.result == BlackJackPlayResult.LOSE }.count()
```
count는 아래 처럼 사용이 가능합니다! 

```suggestion
        val winCount = playerResults.count { it.result == BlackJackPlayResult.LOSE }
```

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackGame.kt:
```diff
@@ -12,8 +12,33 @@ class BlackJackGame(
         return BlackJackGameResult(playerResults, BlackJackDealerResult(winCount, loseCount))
     }
+
+    fun dealerDraw(blackJackDeck: BlackJackDeck): Boolean {
```
해당 함수도 테스트 작성해보면 어떨까요?

gmltkddl1 replied:
![image](https://github.com/user-attachments/assets/a055909a-94ca-4a91-983e-aa002560c95c)

vsh123 reviewed (changes requested):
안녕하세요! 
step4 구현 잘해주셨네요!! 💯 

몇가지 코멘트를 남겼으니 확인부탁드립니다~!

gmltkddl1 committed:
refactor: step4 블랙잭(배팅)

입력시 마다 player 생성하는 방식으로 변경
mutableList 제거
backing property 적용
메소드명 변수명 변경
count 적용
test code 추가

gmltkddl1 committed:
refactor: step4 블랙잭(배팅)

추상화 레벨 맞추기

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackPlayerCards.kt:
```diff
@@ -25,15 +25,19 @@ class BlackJackPlayerCards(private val _cards: MutableList<BlackJackCard> = muta
     }
 
+    fun isBlackJack(): Boolean {
+        return getCardsBestSum().equals(BLACKJACK_NUMBER) && cards.size == 2
```
`==` 비교와 `===` 비교의 차이점을 알아보고 적용해보면 어떨까요?

https://velog.io/@wlsrhkd4023/Kotlin-%EB%8F%99%EB%93%B1%EC%84%B1Equality-%EB%8F%99%EC%9D%BC%EC%84%B1Identify

gmltkddl1 replied:
해당 값은 primitive 타입이라 두개의 차이 없다고 생각했는데 위 코드에서 문제되는 부분이 있나요?

vsh123 replied:
아니요~! 다만 코틀린에서는 equals 비교를 `==` 로 지원하기 때문에 `==` 비교로 통일 시켜주면 좋을 것 같아요!

vsh123 reviewed (changes requested):
희상님 안녕하세요~!

블랙잭 미션 마지막까지 수고하셨습니다!
마지막 질문을 위해 request change로 남겨놓을게요!! 

질문에 대한 답변 완료되면 어프로브하도록 하겠습니다~!

gmltkddl1 committed:
refactor: step4 블랙잭(배팅)

테스트 클래스명 변경
bestsum 피드백 반영

vsh123 reviewed (approved):
안녕하세요!

블랙잭 미션 수행하느라 고생많으셨습니다 :)
이번미션이 많은 도움이 되셨길 바라면서 마무리 하도록 할게요!

고생하셨습니다~!
